src/metabase/models/serialization/base.clj

```diff
@@ -0,0 +1,171 @@
+  is represented by `(or entity_id identity-hash)` of the target entity.
+
+  This is mostly part of the deserialization process, so it's based on the serialized map and not on an IModel instance.
+
+  Default implementation returns an empty vector, so only models that have dependencies need to implement this."
+  :serdes_type)
+
+(defmethod serdes-dependencies :default [_]
+  [])
+
+(p.types/defprotocol+ ISerializable
+  ;;; Serialization side
+  (serialize-all
+    [this user-or-nil]
+    "Serializes all relevant instances of this model.
+    The return value is a reducible sequence of [file-path entity-map] pairs, possibly empty.
+    There can be a single file for all the entities (eg. settings) or one file per entity.
+
+    The default implementation assumes there will be many files:
+    - It calls (serialize-query this user-or-nil) to get the (lazy, reducible) set of entities to serialize.
+    - For each entity, it calls `(serialize-one e)` and expects a `[file-path {...}]` pair.
+    - These pairs are returned as a reducible stream.
+
+    To serialize as a single file, override this to do something else. (Then `serialize-query` and `serialize-one` are
+    never called and can be stubs.)")
+
+  (serialize-query
+    [this user-or-nil]
+    "Performs the select query, possibly filtered, for all the entities of this type that should be serialized.
+    Returns the result of the db/select-reducible.
+    Defaults to a naive `db/select-reducible` for the entire model, ignoring the optional user argument.
+
+    Only called by `serialize-all`, either the default implementation or (optionally) by a custom one.
+    This exists to be an easy override point, when the default `serialize-all` is good but you need to filter
+    the returned set, eg. by dropping archived entities.")
+
+  (serialize-one
+    [this]
+    "Serializes a single entity into a YAML-friendly map, with its filename.
+    Returns a [file-name {entity map...}] pair.
+
+    Default implementation:
+    - Uses `\"$ENTITY_ID.yaml\"` or `\"$IDENTITY_HASH.yaml\"` as the filename.
+    - Returns the entity as a vanilla map with `:id` and any `:foo_at` fields removed.
+    - Adds the field `:type \"Collection\"` (etc.) to specify the model.
+
+    That suffices for a few simple entities, but most entities will need to override this.
+    They should follow the pattern of dropping timestamps and numeric database IDs, and including the `:type`.")
+
+  ;;; Deserialization side.
+  (deserialize-file
+    [this new-map local]
+    "Given the model, the file contents as an EDN map, and any corresponding entity already existing in the database,
+    this performs whatever steps are necessary to update or insert the incoming entity.
+    Called for each **file**, which usually means for each entity. For single-file output like for settings, this is
+    called once for the whole file.
+    Defaults to [[deserialize-file-plus]] with [[identity]] as the inner transformation.")
```

Comment:
I like the fact that this is decoupled from the actual format of the files themselves (i.e. YAML) and the implementation doesn't have to worry about that. I suggest taking the abstraction a bit further tho. It seems a little weird that a model's implementation has to concern itself with whether or not some specific file contained a single object or multiple objects. It seems like that's some thing that can be handled in the shared code and the actual implementations could be `deserialize-one` or `deserialize-many` instead. `deserialize-many` could just repeatedly call `deserialize-one` by default

Reply:
I've made the refactoring you suggested; the main deserialization now has no connection to files at all, just an abstracted `ingestion` surface.